sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala

```diff
@@ -381,7 +381,7 @@ case class FileSourceScanExec(
       fsRelation: HadoopFsRelation): RDD[InternalRow] = {
     logInfo(s"Planning with ${bucketSpec.numBuckets} buckets")
     val filesGroupedToBuckets =
-      selectedPartitions.flatMap { p =>
+      selectedPartitions.toArray.flatMap { p =>
```

Comment:
Does it make sense to apply the same change to `selectedPartitions.flatMap` in `createNonBucketedReadRDD(...)`?

Reply:
Don't you want to put the `.toArray` much later here, like line 404?

Reply:
Ah, although I thought `toArray.flatMap` was a kind of optimization. However, I cannot find `toArray.flatMap` in any other place.   
It looks good to keep the original code.

Reply:
Yeah, I think just ensure the final result in the case class is an array. I'm not clear it would be faster to convert to an array and then maybe implicitly convert back to another type along the way, etc.

Reply:
`@transient private lazy val selectedPartitions: Seq[PartitionDirectory]` is returning type of `Seq`. Since it's `lazy val` and used many times in the life cycle, why not we make it `@transient private lazy val selectedPartitions: Array[PartitionDirectory]`?

Reply:
That's probably fine if we need it to be an Array, or at least an IndexedSeq

Reply:
@dbtsai I like your suggestion but it requires that we either change the return type of `FileIndex.listFiles`  or convert the value returned by `relation.location.listFiles` to an array, which doesn't save us anything.  Changing the return type of `FileIndex.listFiles` is a bigger change that I would want to make at this point, given my relative inexperience with the code.

@srowen, 
```
 val filePartitions: Seq[FilePartition] = Seq.tabulate(bucketSpec.numBuckets) { bucketId =>
      FilePartition(bucketId, prunedFilesGroupedToBuckets.getOrElse(bucketId, Nil).toArray)
    }.toArray
```
doesn't compile 
```
/spark/sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala:404: polymorphic expression cannot be instantiated to expected type;
[error]  found   : [B >: org.apache.spark.sql.execution.datasources.PartitionedFile]Array[B]
[error]  required: Seq[org.apache.spark.sql.execution.datasources.PartitionedFile]
[error]       FilePartition(bucketId, prunedFilesGroupedToBuckets.getOrElse(bucketId, Nil).toArray)
```
Scala isn't my first language so if you can help me with a way to satisfy the compiler, I'll be happy to make the change.

Reply:
I also checked whether the return type can be changed to `Array[]`. I also realized that it is not a small change...

Reply:
`filePartitions` doesn't need to change. You declare it as a Seq but that's incompatible with the second `.toArray`. You just want the first `.toArray`. That works for me.

Reply:
Could `filePartitions` be a large list and run into the same serialization issue if we don't convert it into an array? Doesn't look likely, but I'd like to be sure. 
Also, @DBTsai suggested `  @transient private lazy val selectedPartitions: Array[PartitionDirectory] = {`. Do we want to do that?